File: src/core/types/map.ts
// ─────────────────────────────────────────────────────────────────────────
// 자유 방향그래프 맵 스키마 (디자이너 데이터 계약, 7장). 에디터가 JSON으로 저작.
// 엔진(core/run/graph.ts)이 해석 — 좌표암시 간선(구 genMap) 폐기, 명시적 방향 간선.
// 이동=간선 방향, 복귀 불가. clear 노드 진입=층 종료. 런=층 그래프(clear.toFloor 링크, 분기 가능).
// ─────────────────────────────────────────────────────────────────────────
import type { NodeType, Pos } from "./content.ts";
import type { PassiveRule } from "./passives.ts";

// ── 노드 레이어 (NODE-DESIGN Phase A) ─────────────────────────────────────
// 노드 = 타입 코어 + 부착 레이어. 레이어 = 진입/완료 슬롯에 순서 실행하는 효과.
// Phase A 슬라이스1 = 즉시(데코레이터) 레이어만. 상호작용 모듈(combat 등)은 A2~.
// 레이어 = 데코레이터(즉시 실행) + 상호작용 모듈(완료까지 블록). 해석: 데코=helpers.runInstantLayers, 상호작용=run/layers.ts 시퀀서.
/** 즉시 실행 데코레이터 레이어 — 어느 노드든 부착 가능(전역). */
export type DecoratorLayer =
  | { kind: "gold"; amount: number } // 골드 ±(0 미만 클램프)
  | { kind: "heal"; pct: number; revive?: boolean } // 파티 회복(maxHp 비율, revive=전투불능 부활)
  | { kind: "grantStatus"; charId?: string; statusId: string; stacks: number; duration: number } // 다음 전투 계승(charId 없으면 전원)
  | { kind: "text"; text: string }; // 로그/대사(컷신 뷰 강화는 Phase C)
/** 상호작용 레이어 — 완료까지 블록(phase 전환). combat·reward(B). shop/event는 후속. */
export type InteractiveLayer =
  | { kind: "combat"; roster?: { charId: string; pos: Pos }[]; rosterPreset?: string; boss?: boolean; rules?: PassiveRule[] } // 적: roster>rosterPreset>battle. 보스=진형보너스. rules=이 전투의 트리거 룰(페이즈별 대사·개입, Phase C/E4)
  | { kind: "reward" } // 보상 3택1(genRewards) — 플레이어 선택까지 블록. treasure 노드 = core:[reward]
  | { kind: "shop" } // 상점 진열(generateShop) — leaveShop까지 블록. 스타터는 DI 등록(run.ts)
  | { kind: "event" }; // 인카운터 추첨(랜덤) — chooseEncounterOption까지 블록. 노드별 저작은 Phase D
/** 노드 레이어 — onEnter/onResolve는 데코만(즉시), core는 데코+상호작용 혼합(순서 실행). */
export type Layer = DecoratorLayer | InteractiveLayer;
export type LayerKind = Layer["kind"];
/** 노드의 부착 레이어 — 슬롯별 순서 리스트. onEnter=진입 직후, onResolve=노드 완료 시. */
export interface NodeLayers {
  onEnter?: DecoratorLayer[];
  onResolve?: DecoratorLayer[];
}

/** 맵 노드 — q,r은 에디터 캔버스/렌더 좌표(위상 아님; 위상은 edges가 정의). */
export interface MapNode {
  id: string;
  type: NodeType; // start=입장, clear=목표 마커, 그 외=노드별 해소(전투/상점/휴식/인카운터)
  q: number;
  r: number;
  /** clear 노드 전용: 진입 시 갈 다음 층 id. 없으면 그 클리어 = 런 승리(종료). 여러 clear = 분기. */
  toFloor?: string;
  /** 전투 노드 전용: 적 구성 override. 없으면 타입 기본(NODE_ROSTERS[type])을 쓴다. */
  roster?: { charId: string; pos: Pos }[];
  /** 표시 라벨(선택). 노드 위에 표기 — 같은 타입 노드를 구분(예: "두목 호위대"). */
  label?: string;
  /** 부착 레이어(선택, Phase A). 없으면 순수 타입 코어 동작(기존과 동일). */
  layers?: NodeLayers;
  /** 코어 시퀀스(선택, Phase A2). 있으면 타입 분기 대신 이 레이어들을 순서 실행(combat 웨이브 등). 트리거 룰은 각 combat 레이어가 소유. */
  core?: Layer[];
}

/** 방향 있는 간선 — from에서 to로만 전진(복귀 불가). */
export interface MapEdge {
  from: string;
  to: string;
}

/** 한 층 = 방향그래프. entry에서 출발, 어느 clear 노드든 진입하면 층 종료. */
export interface FloorDef {
  id: string;
  name?: string;
  entryNodeId: string;
  nodes: MapNode[];
  edges: MapEdge[];
}

/** 런 = 층 그래프(floors 집합, clear.toFloor로 연결) + 시작 파티 + 모드 설정. (구 GameMode 흡수) */
export interface RunDef {
  id: string;
  name: string;
  desc?: string;
  useMastery: boolean;
  entryFloorId: string; // 시작 층 id
  roster: { charId: string; pos: Pos }[];
  floors: FloorDef[]; // 순서 무의미 — 탐색은 id(entryFloorId·clear.toFloor)로
}

File: engine/spr-types/src/map.rs
//! 런 맵 타입 (TS `types/map.ts`의 그래프 부분). 레이어/core 본문은 시퀀서 슬라이스(P2-4)서 도입(현재 serde 무시).
use crate::data::Placement;
use crate::passives::PassiveRule;
use serde::{Deserialize, Serialize};

/// 노드 트리거 룰 소유자(화자/기준).
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RuleOwner {
    pub side: String, // ally|enemy
    #[serde(rename = "charId")]
    pub char_id: String,
}

/// 노드 트리거 룰 = PassiveRule + owner. owner의 side+charId 유닛에 주입(self=그 개체). TS NodeRule.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct NodeRule {
    #[serde(flatten)]
    pub rule: PassiveRule,
    #[serde(default)]
    pub owner: Option<RuleOwner>,
}

/// 상점 진열 저작(판별 `kind`). TS ShopOfferDef.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "kind")]
pub enum ShopOfferDef {
    #[serde(rename = "buyItem")]
    BuyItem { #[serde(rename = "itemId")] item_id: String, cost: i64 },
    #[serde(rename = "heal")]
    Heal { pct: i64, cost: i64 },
    #[serde(rename = "learn")]
    Learn { #[serde(rename = "charId")] char_id: String, #[serde(rename = "skillId")] skill_id: String, cost: i64 },
}

/// 노드 레이어(판별 `kind`) — 데코(즉시) + 상호작용(블록). TS Layer.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(tag = "kind")]
pub enum Layer {
    // 데코레이터(즉시)
    #[serde(rename = "gold")]
    Gold { amount: i64 },
    #[serde(rename = "heal")]
    Heal { pct: i64, #[serde(default)] revive: Option<bool> },
    #[serde(rename = "grantStatus")]
    GrantStatus { #[serde(rename = "charId", default)] char_id: Option<String>, #[serde(rename = "statusId")] status_id: String, stacks: i64, duration: i64 },
    #[serde(rename = "text")]
    Text { text: String },
    /// 파티 변동(즉시) — 런 중 합류/이탈(스토리). add=charId 신규 합류(루트 직업·숙련0·빈 슬롯 배치), remove=charId 이탈.
    #[serde(rename = "partyChange")]
    PartyChange {
        #[serde(default)]
        add: Option<Vec<String>>,
        #[serde(default)]
        remove: Option<Vec<String>>,
    },
    // 상호작용(블록)
    #[serde(rename = "combat")]
    Combat {
        #[serde(default)] roster: Option<Vec<Placement>>,
        #[serde(default)] boss: bool,
        #[serde(default)] rules: Option<Vec<NodeRule>>,
        /// 자원 게이지 조건부 전투 시작 모디파이어(R1) — 충족 시 side 전원에 상태 주입(민심高→아군 버프 / 심리전→적 fear).
        #[serde(rename = "resourceMods", default)] resource_mods: Option<Vec<ResourceMod>>,
    },
    #[serde(rename = "reward")]
    Reward { #[serde(default)] tier: Option<i64> },
    #[serde(rename = "shop")]
    Shop { #[serde(default)] offers: Option<Vec<ShopOfferDef>>, #[serde(rename = "keepGenerated", default)] keep_generated: Option<bool> },
    #[serde(rename = "event")]
    Event { #[serde(default)] event: Option<EncounterEvent> },
    /// 전직(4.7) 상호작용 레이어 — 전직 가능 파티원 중 최대 `max`명 전직. 전직노드(2~3)·쉼터(1) 작곡용.
    #[serde(rename = "classChange")]
    ClassChange {
        #[serde(default = "default_one")]
        max: i64,
    },
    /// 런 자원 변경(R1, 즉시) — id 자원을 delta만큼 가감(min/max 클램프). event 선택지·노드 onResolve용.
    #[serde(rename = "resource")]
    Resource {
        id: String,
        delta: i64,
    },
}

/// 런 자원 정의(R1) — 민심·명예·토사구팽 등 런-영속 명명 자원. RunDef.resources.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ResourceDef {
    pub id: String,
    pub name: String,
    pub min: i64,
    pub max: i64,
    pub initial: i64,
    #[serde(default)]
    pub icon: Option<String>,
}

/// 자원 임계 비교(R1) — resourceMods·EncounterChoice.requires 공용. cmp = gte|lte|gt|lt|eq.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ResourceReq {
    #[serde(rename = "resourceId")]
    pub resource_id: String,
    pub cmp: String,
    pub value: i64,
}

/// 전투 시작 자원 모디파이어(R1) — 자원이 임계를 충족하면 side 전원에 상태 주입.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ResourceMod {
    #[serde(rename = "resourceId")]
    pub resource_id: String,
    pub cmp: String,
    pub value: i64,
    pub side: String, // ally|enemy
    #[serde(rename = "statusId")]
    pub status_id: String,
    pub stacks: i64,
    pub duration: i64,
}

/// cmp 평가(R1 공용) — gte|lte|gt|lt|eq.
pub fn cmp_ok(cmp: &str, lhs: i64, rhs: i64) -> bool {
    match cmp {
        "gte" => lhs >= rhs,
        "lte" => lhs <= rhs,
        "gt" => lhs > rhs,
        "lt" => lhs < rhs,
        "eq" => lhs == rhs,
        _ => false,
    }
}

fn default_one() -> i64 {
    1
}

/// 노드 부착 레이어 슬롯(onEnter/onResolve=데코). TS NodeLayers.
#[derive(Debug, Clone, Deserialize, Serialize, Default)]
pub struct NodeLayers {
    #[serde(rename = "onEnter", default)]
    pub on_enter: Option<Vec<Layer>>,
    #[serde(rename = "onResolve", default)]
    pub on_resolve: Option<Vec<Layer>>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct MapNode {
    pub id: String,
    #[serde(rename = "type")]
    pub node_type: String, // start|battle|elite|shop|encounter|rest|boss|clear
    pub q: i64,
    pub r: i64,
    #[serde(rename = "toFloor", default)]
    pub to_floor: Option<String>, // clear 노드의 다음 층(분기). 없으면 승리 클리어
    #[serde(default)]
    pub label: Option<String>,
    #[serde(default)]
    pub layers: Option<NodeLayers>,
    #[serde(default)]
    pub core: Option<Vec<Layer>>,
}

#[derive(Debug, Clone, Deserialize, serde::Serialize)]
pub struct MapEdge {
    pub from: String,
    pub to: String,
}

/// 인카운터 선택 결과(판별 `kind`). TS EncounterOutcome.
#[derive(Debug, Clone, Deserialize, serde::Serialize)]
#[serde(tag = "kind")]
pub enum EncounterOutcome {
    #[serde(rename = "heal")]
    Heal { pct: i64 },
    #[serde(rename = "hurt")]
    Hurt { pct: i64 },
    #[serde(rename = "gold")]
    Gold { amount: i64 },
    #[serde(rename = "upgradeRandom")]
    UpgradeRandom,
    #[serde(rename = "learnUniversal")]
    LearnUniversal,
    #[serde(rename = "nothing")]
    Nothing,
    /// 런 자원 가감(R1) — id 자원 delta.
    #[serde(rename = "resource")]
    Resource { id: String, delta: i64 },
}

#[derive(Debug, Clone, Deserialize, serde::Serialize)]
pub struct Gamble {
    pub chance: i64,
    pub win: EncounterOutcome,
    pub lose: EncounterOutcome,
}

#[derive(Debug, Clone, Deserialize, serde::Serialize)]
pub struct EncounterChoice {
    pub id: String,
    pub label: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<EncounterOutcome>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gamble: Option<Gamble>,
    /// 자원 게이팅(R1) — 충족해야 선택 가능(예: 민심 gte 60). 미충족=비활성.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub requires: Option<ResourceReq>,
}

#[derive(Debug, Clone, Deserialize, serde::Serialize)]
pub struct EncounterEvent {
    pub id: String,
    pub title: String,
    pub text: String,
    pub choices: Vec<EncounterChoice>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct FloorDef {
    pub id: String,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(rename = "entryNodeId")]
    pub entry_node_id: String,
    pub nodes: Vec<MapNode>,
    pub edges: Vec<MapEdge>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RunDef {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub desc: Option<String>,
    /// 모드(8.8): campaign 등. 셸이 분류·노출. 엔진 미해석(전체 선언 규약).
    #[serde(default)]
    pub mode: Option<String>,
    #[serde(rename = "useMastery")]
    pub use_mastery: bool,
    #[serde(rename = "entryFloorId")]
    pub entry_floor_id: String,
    pub roster: Vec<Placement>,
    pub floors: Vec<FloorDef>,
    /// 런 자원 게이지(R1) — 민심·명예 등 런-영속 자원 정의. 미지정=자원 없는 런.
    #[serde(default)]
    pub resources: Vec<ResourceDef>,
}
